waspc/src/Wasp/Generator/ServerGenerator/EmailG.hs

```diff
@@ -0,0 +1,143 @@
+genCoreHelpers email = return $ C.mkTmplFdWithDstAndData srcPath dstPath (Just tmplData)
+  where
+    srcPath = C.srcDirInServerTemplatesDir </> [relfile|email/core/helpers.ts|]
+    dstPath = C.serverSrcDirInServerRootDir </> [relfile|email/core/helpers.ts|]
+    tmplData =
+      object
+        [ "senderDefaults"
+            .= object
+              [ "email" .= AS.EmailSender.email emailFrom,
+                "title" .= title,
+                "isTitleDefined" .= isJust title
+              ]
+        ]
+    emailFrom = AS.EmailSender.defaultFrom email
+    title = AS.EmailSender.title emailFrom
+
+genSmtp :: EmailSender -> Generator [FileDraft]
+genSmtp email =
+  case AS.EmailSender.provider email of
+    AS.EmailSender.SMTP ->
+      sequence
+        [ copyTmplFile [relfile|email/core/providers/smtp.ts|]
+        ]
+    _ -> return []
+
+genSendGrid :: EmailSender -> Generator [FileDraft]
+genSendGrid email =
+  case AS.EmailSender.provider email of
+    AS.EmailSender.SendGrid ->
+      sequence
+        [ copyTmplFile [relfile|email/core/providers/sendgrid.ts|]
+        ]
+    _ -> return []
+
+genMailgun :: EmailSender -> Generator [FileDraft]
+genMailgun email =
+  case AS.EmailSender.provider email of
+    AS.EmailSender.Mailgun ->
+      sequence
+        [ copyTmplFile [relfile|email/core/providers/mailgun.ts|]
+        ]
+    _ -> return []
+
+getEmailProvidersJson :: EmailSender -> Aeson.Value
+getEmailProvidersJson email =
+  object
+    [ "isSmtpProviderUsed" .= isSmtpProviderUsed,
+      "isSendGridProviderUsed" .= isSendGridProviderUsed,
+      "isMailgunProviderUsed" .= isMailgunProviderUsed
+    ]
+  where
+    isSmtpProviderUsed = provider == AS.EmailSender.SMTP
+    isSendGridProviderUsed = provider == AS.EmailSender.SendGrid
+    isMailgunProviderUsed = provider == AS.EmailSender.Mailgun
+    provider = AS.EmailSender.provider email
+
+copyTmplFile :: Path' (Rel C.ServerTemplatesSrcDir) File' -> Generator FileDraft
+copyTmplFile = return . C.mkSrcTmplFd
+
+-- Dependencies
+
+nodeMailerVersionRange :: SV.Range
+nodeMailerVersionRange = SV.Range [SV.backwardsCompatibleWith (SV.Version 6 9 1)]
+
+nodeMailerDependency :: AS.Dependency.Dependency
+nodeMailerDependency = AS.Dependency.make ("nodemailer", show nodeMailerVersionRange)
+
+sendGridVersionRange :: SV.Range
+sendGridVersionRange = SV.Range [SV.backwardsCompatibleWith (SV.Version 7 7 0)]
+
+sendGridDependency :: AS.Dependency.Dependency
+sendGridDependency = AS.Dependency.make ("@sendgrid/mail", show sendGridVersionRange)
+
+mailgunVersionRange :: SV.Range
+mailgunVersionRange = SV.Range [SV.backwardsCompatibleWith (SV.Version 0 5 1)]
+
+mailgunDependency :: AS.Dependency.Dependency
+mailgunDependency = AS.Dependency.make ("ts-mailgun", show mailgunVersionRange)
+
+depsRequiredByEmail :: AppSpec -> [AS.Dependency.Dependency]
+depsRequiredByEmail spec =
+  concat
+    [ [nodeMailerDependency | provider == Just AS.EmailSender.SMTP],
```

Comment:
Here is a thought:
So you have a bunch of function here in this file, and a lot of them do checks to figure out which provider is it, and then based on that, provide correct data/logic/result.

So in bunch of places we have "if smtp then X else if sendgrid then Y else if mailgun then Z" -> I am paraphrasing, but something like that.

That is not bad, it is pretty normal for functional programming, but when reading this file it did feel a bit dissipated maybe, like it is hard to put together a picture of what we end up with if we do smtp, what we end up with if we do sendgrid, ... .

Maybe, it would make sense to reorganize it a bit, and have one central function that goes smth like `getEmailSenderInfo` (ok name sucks, and maybe that indicates that this is the wrong approach) that has one big `case` and then returns all the stuff for that provider -> its depenencies, its core stuff, and maybe there is something more?

I am not sure if this is better than what yo have right now, but it sounds potentially interesting, so I think it is maybe worth a quick try and if it looks better, great.

Reply:
I've tried to do that in `Wasp.Generator.ServerGenerator.EmailSender.Providers`, grouping of all relevant data for each of the providers.